File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque
class Solution:
    def sameTree(self, left: Optional[TreeNode], right: Optional[TreeNode]) -> bool:
        if not left and not right:
            return True
        if left and right:
            return left.val == right.val and self.sameTree(left.left, right.left) and self.sameTree(left.right, right.right)
        return False

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root is None and subRoot is None:
            return True
        if root and subRoot:
            return (root.val == subRoot.val and self.sameTree(root, subRoot)) or (self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot))
        return False

File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn validate(node: Option<Rc<RefCell<TreeNode>>>, low: i32, high: i32) -> bool {
        // The current node values must be between the low and the high
        if node.is_none() {
            return true
        }
        let node = node.unwrap();
        let node_ref = node.borrow();
        if low < node_ref.val && node_ref.val < high {
            return Solution::validate(node_ref.left.clone(), low, node_ref.val) && Solution::validate(node_ref.right.clone(), node_ref.val, high)
        }

        return false
    }

    pub fn is_valid_bst(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if root.is_none() {
            return true
        }
        Solution::validate(root, i32::MIN, i32::MAX)
    }
}
